python/ray/log.py

```diff
@@ -0,0 +1,158 @@
+        super().__init__(*args, **kwargs)
+        self.global_worker_mode = None
+
+    def filter(self, record: logging.LogRecord) -> bool:
+        """Add context information to the log record.
+
+        This filter adds a package name from where the message was generated as
+        well as the worker IP address, if applicable.
+
+        Args:
+            record: Record to be filtered
+
+        Returns:
+            True if the record is to be logged, False otherwise. (This filter only
+            adds context, so records are always logged.)
+        """
+        match = self.logger_regex.search(record.name)
+        if match and match["subpackage"] in self.package_message_names:
+            record.package = f"[Ray {self.package_message_names[match['subpackage']]}]"
+        else:
+            record.package = ""
+
+        # Lazily load ray when this is called to avoid circular import
+        import ray
+
+        if hasattr(ray, "_private"):
+            if self.global_worker_mode is None:
+                self.global_worker_mode = ray._private.worker.global_worker.mode
+
+            if (
+                self.global_worker_mode
+                and self.global_worker_mode == ray._private.worker.WORKER_MODE
+            ):
+                record.node_ip_address = (
+                    ray.runtime_context.get_runtime_context().worker.node_ip_address
+                )
+
+        return True
+
+
+class AnsiStripFormatter(logging.Formatter):
+    """Formatter which strips ANSI escape codes from log messages.
+
+    These escape codes produce noise in output files and conflict with logging,
+    and so are stripped here.
+    """
+
+    strip_ansi_regex = re.compile(r"\x1b\[[0-9;]*m")
+
+    def format(self, record: logging.LogRecord) -> str:
+        return re.sub(self.strip_ansi_regex, "", super().format(record))
+
+
+class PlainRayHandler(logging.StreamHandler):
+    """A plain log handler."""
+
+    def __init__(self):
+        super().__init__(stream=sys.stdout)
+        self.plain_handler = logging.StreamHandler()
+        self.plain_handler.level = self.level
+        self.plain_handler.formatter = logging.Formatter(fmt="%(message)s")
+
+    def emit(self, record: logging.LogRecord):
+        """Emit the log message.
+
+        If this is a worker, serialize the record and send it to the driver via TCP.
+        If this is the driver, emit the message using the appropriate console handler.
+
+        Args:
+            record: Log record to be emitted
+        """
+        import ray
+
+        if (
+            hasattr(ray, "_private")
+            and ray._private.worker.global_worker.mode
+            == ray._private.worker.WORKER_MODE
+        ):
+            self.plain_handler.emit(record)
+        else:
+            logging.StreamHandler.emit(self, record)
+
+
+logger_initialized = False
+
+
+def generate_logging_config():
+    """Generate the default Ray logging configuration."""
+    global logger_initialized
```

Comment:
Maybe we need to lock this?

Reply:
Possibly? It's only called once, currently, at `ray` module initialization time. The previous `call_once` decorator effectively implemented a lock, but I can add one here too.

Reply:
yep can you add a lock just in case? should be a very simple to add one